New actions to list and remove images

kwmonroe opened the pull request:
- Address https://bugs.launchpad.net/layer-docker-registry/+bug/1808200 with an action that lets operators remove registry images.  This is effectively a wrapper around `docker rmi`.
- Let the people list images while we're at it.  This is effectively `docker images`.

These actions have been released in charm rev 125, currently in the beta channel:
https://jujucharms.com/u/containers/docker-registry/125

tvansteenburgh commented on actions/images:
```diff
@@ -0,0 +1,38 @@
+
+def fail(msg):
+    hookenv.action_fail(msg)
+    sys.exit()
```
Technically this should be `sys.exit(msg)` or `sys.exit(1)`. As is, juju will report that this action completed, when it actually failed.

kwmonroe replied:
When using `sys.exit($non-zero)`, juju shows a generic return code error, and the overall action rc is still 0:
```bash
## with sys.exit(1), unhelpful message
$ juju run-action --wait docker-registry/0 images options=-badtim
unit-docker-registry-0:
  id: 964b7a31-3080-4853-82c0-a002c99f3cf0
  message: exit status 1
  status: failed
  timing:
    completed: 2019-04-26 21:14:27 +0000 UTC
    enqueued: 2019-04-26 21:14:26 +0000 UTC
    started: 2019-04-26 21:14:26 +0000 UTC
  unit: docker-registry/0
$ echo $?
0

## with sys.exit('this is fine'), unhelpful message; supplied msg goes to debug log
$ juju run-action --wait docker-registry/0 images options=-badtim
unit-docker-registry-0:
  id: 0f743990-50a7-4f95-84fd-9729b977ec90
  message: exit status 1
  status: failed
  timing:
    completed: 2019-04-26 21:13:31 +0000 UTC
    enqueued: 2019-04-26 21:13:29 +0000 UTC
    started: 2019-04-26 21:13:30 +0000 UTC
  unit: docker-registry/0
$ echo $?
0
```

Whereas `sys.exit($zero)` allows our helpful `action_fail` message to come through:
```
$ juju run-action --wait docker-registry/0 images options=-badtim
unit-docker-registry-0:
  id: 811696ab-75c8-4dcf-8de7-e291dbfa625f
  message: |
    "docker images -badtim" failed with:
    unknown shorthand flag: 'b' in -badtim
    See 'docker images --help'.
  status: failed
  timing:
    completed: 2019-04-26 21:17:20 +0000 UTC
    enqueued: 2019-04-26 21:17:16 +0000 UTC
    started: 2019-04-26 21:17:19 +0000 UTC
  unit: docker-registry/0
$ echo $?
0
```

tvansteenburgh replied:
Huh, TIL.

tvansteenburgh commented on actions/images:
```diff
@@ -0,0 +1,38 @@
+cmd = ['docker', 'images']
+options = hookenv.action_get('options')
+if options:
+    cmd.extend(options.split())
```
I always use `shlex.split()` for command-line strings.

kwmonroe replied:
Yikes.  I even touted the `options` param as 'gimme anything you'd give to the shell'!  Nice catch.

tvansteenburgh commented on actions/images:
```diff
@@ -0,0 +1,38 @@
+except CalledProcessError as e:
+    fail("{} failed with: {}".format(e.cmd, e.output.decode('utf-8')))
+else:
+    hookenv.action_set({'outcome': 'success'})
```
I feel like this is redundant - Juju adds a `status: [pending|completed|failed]` already.

kwmonroe replied:
I think once upon a time, Juju action output didn't have the `status` line, and so you'd see `timing: completed` even on failed jobs and perhaps miss the fact it failed.

Even if I just made that up, your point stands. `outcome` is redundant with `status` in the output.

kwmonroe commented:
Addressed review comments (🍰  @tvansteenburgh) and updated a bit of action output.  Ready for re-review.

```bash
## good listing
$ juju run-action --wait docker-registry/0 images repository=ubuntu
unit-docker-registry-0:
  id: f98c5411-dcc0-4cff-88bf-f558e2c34739
  results:
    output: |
      REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
      ubuntu              18.04               94e814e2efa8        6 weeks ago         88.9MB
  status: completed
  timing:
    completed: 2019-04-26 21:33:52 +0000 UTC
    enqueued: 2019-04-26 21:33:49 +0000 UTC
    started: 2019-04-26 21:33:52 +0000 UTC
  unit: docker-registry/0

## bad listing
$ juju run-action --wait docker-registry/0 images options=-ouch
unit-docker-registry-0:
  id: 1c98bc5c-d02d-45bc-8a16-1a82dce7c69f
  message: |
    "docker images -ouch" failed with:
    unknown shorthand flag: 'o' in -ouch
    See 'docker images --help'.
  status: failed
  timing:
    completed: 2019-04-26 21:34:14 +0000 UTC
    enqueued: 2019-04-26 21:34:09 +0000 UTC
    started: 2019-04-26 21:34:13 +0000 UTC
  unit: docker-registry/0

## good remove
$ juju run-action --wait docker-registry/0 rmi image=ubuntu:18.04
unit-docker-registry-0:
  id: 9dd49ac6-9dd6-4374-8028-f5c5241b8bcd
  results:
    output: |
      Untagged: ubuntu:18.04
      Untagged: ubuntu@sha256:017eef0b616011647b269b5c65826e2e2ebddbe5d1f8c1e56b3599fb14fabec8
  status: completed
  timing:
    completed: 2019-04-26 21:34:32 +0000 UTC
    enqueued: 2019-04-26 21:34:30 +0000 UTC
    started: 2019-04-26 21:34:32 +0000 UTC
  unit: docker-registry/0

## bad remove
$ juju run-action --wait docker-registry/0 rmi image=ubuntu:18.04
unit-docker-registry-0:
  id: 80ff7494-ed1a-4da8-8f97-3fc371e68f73
  message: |
    "docker rmi ubuntu:18.04" failed with:
    Error: No such image: ubuntu:18.04
  status: failed
  timing:
    completed: 2019-04-26 21:36:38 +0000 UTC
    enqueued: 2019-04-26 21:36:37 +0000 UTC
    started: 2019-04-26 21:36:37 +0000 UTC
  unit: docker-registry/0

## it's gone
$ juju run-action --wait docker-registry/0 images repository=ubuntu
unit-docker-registry-0:
  id: 3c8fc59d-d4d6-4eb9-8393-28734b567605
  results:
    output: |
      REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
  status: completed
  timing:
    completed: 2019-04-26 21:34:46 +0000 UTC
    enqueued: 2019-04-26 21:34:45 +0000 UTC
    started: 2019-04-26 21:34:45 +0000 UTC
  unit: docker-registry/0
```
